### DISCUSSION TITLE
Any self-contained terminal?

### DISCUSSION THREAD

**Discussion by iliakan** [sourcelair/xterm.js#4414] [Q&A] [answered]
Hello,

Are there any good addons to make a self-contained terminal?

In other words, a browser-only (without a backend) terminal with history, keys support (arrows, backspace etc), maybe other features.

I saw some examples, like https://github.com/wavesoft/local-echo, but they appear to be abandoned and somewhat buggy.

What you'd suggest to use?

**Comment by jerch** [comment, marked as answer]
As far as I am aware there is no ready-to-go solution w'o flaws for pure in-browser handling with xterm.js. There mainly 2 big issues behind it - first the missing pty implementation, and second a shell variant resembling typical key stroke pattern one might expect to work.
You are pretty much on your own here beside those solutions pointed out by @Tyriar here https://github.com/xtermjs/xterm.js/discussions/4404#discussioncomment-4905959 or the one you mentioned above.

I started https://github.com/jerch/browser-fakepty out of curiosity, how far I'd get with that in pure browser realms, but it is nothing more than a toy experiment atm. While one can get somewhere with that approach, it still needs serious coding efforts to get the basics straight. As I had no own serious use case for it I stopped development when more and POSIX abstraction layers got pulled in partially resembling given kernel interfaces, while drawbacks like being bound to coop-style only cannot be solved easily (esp. emulating blocking IO with native languages compiled to wasm).

**Reply by iliakan** [reply]
Thank you for the in-depth answer 😀

What I need is basically a CLI interface. 

Not a full-fledged terminal for ncurses etc. 
Not a real unix-compatible terminal with all its commands and signals.

Instead, for an online-tutorial that involves CLI, there should be an in-browser CLI environment to type commands, track history etc, show ASCII-colored output. 

A feature-cut version, if I might say ;)

**Reply by jerch** [reply]
I think with some tweaking you can use local-echo for that? Thats at least what wasm-terminal uses (+ some dash wasm variant, you can test it here: https://webassembly.sh/).

**Reply by iliakan** [reply]
When you say `local-echo`, do you mean https://github.com/wavesoft/local-echo?

It has a bug in its core machinery ;(
Quite often one can see flickering when they type or delete a character.
I suppose that's because for most operations it clears the terminal line and re-fills it again. 

Fixing this appears to be the same as rewriting (hope I'm mistaken, so it seemed), because many places in code rely on such behavior.

**Reply by jerch** [reply]
> When you say `local-echo`, do you mean https://github.com/wavesoft/local-echo?

Yes.

> It has a bug in its core machinery ;( Quite often one can see flickering when they type or delete a character. I suppose that's because for most operations it clears the terminal line and re-fills it again.

Idk the source code of local-echo, but in theory update flickering in xterm.js can be avoided by applying all changes at once in one final `term.write` call. Maybe try to stack up the changes and call `write` only once?

**Reply by iliakan** [reply]
Yes, it kind of works, I replaced all `term.write` calls with a queue that flushes on `requestAnimationFrame`. 
Now I have another bug there, with colored prompts (text length incorrect), will need to fix as well.

Is there any reason to prefer xtermjs over a fully custom implementation with a div and textarea?
What are the benefits (surely there are some)?

The purpose is to emulate CLI, but there's no interaction and graphics. Just sending commands, printing response.

**Reply by jerch** [reply]
> Is there any reason to prefer xtermjs over a fully custom implementation with a div and textarea?
What are the benefits (surely there are some)?

From appside a terminal (emulator) is just a way of a decoupled screen representation for the app, with an IO interface orchestrated via escape sequences and a text grid promise with cursor mechanics. The interface is quite convoluted mostly for historical reasons, when terminals were real hardware devices only offering a single IO byte line.
Whether you gonna need that in your app I cannot tell. You might get faster to something usable with your own output system, but might introduce hard to overcome issues at other ends (eg. cursor repositioning + overwrite is quite easy in a terminal, but will get more complex with your own coloring output system).
And for OS cli apps a terminal driven interface is the minimal denominator working almost everywhere. So a terminal is mostly about whats already there on systems.

If you are sure to stay in the browser forever with your app, maybe jquery.terminal is the better choice (it still would give you cursor mechs but w'o the historical escape sequence bloat).
